oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenIntrospectionEndpointFilter.java

```diff
@@ -0,0 +1,450 @@
+			builder.clientID(new ClientID(clientAuthentication.getRegisteredClient().getClientId()));
+			Optional.ofNullable(oauthToken.getIssuedAt()).map(Date::from).ifPresent(builder::issueTime);
+			Optional.ofNullable(oauthToken.getExpiresAt()).map(Date::from).ifPresent(builder::expirationTime);
+			TokenToIntrospectionResponseFieldsMapper.extractFromToken(oauthToken, builder);
+			return builder.build();
+		} catch (InvalidTokenException exception) {
+			TokenIntrospectionSuccessResponse.Builder builder = new TokenIntrospectionSuccessResponse.Builder(false);
+			return builder.build();
+		}
+	}
+
+	private void sendErrorResponse(HttpServletResponse response, OAuth2Error error) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		httpResponse.setStatusCode(HttpStatus.BAD_REQUEST);
+		this.errorHttpResponseConverter.write(error, null, httpResponse);
+	}
+
+	private Optional<OAuth2Authorization> findAuthorizationByToken(String token, Optional<TokenType> tokenTypeHint) {
+		List<TokenType> supportedTokenTypes = new ArrayList<>(
+				Arrays.asList(TokenType.ACCESS_TOKEN, TokenType.REFRESH_TOKEN, TokenType.ID_TOKEN));
+		tokenTypeHint.ifPresent(tType -> {
+			if (supportedTokenTypes.remove(tType))
+				supportedTokenTypes.add(0, tType);
+		});
+		for (TokenType tokenType : supportedTokenTypes) {
+			OAuth2Authorization authorization = this.authorizationService.findByToken(token, tokenType);
+			if (authorization != null) {
+				return Optional.of(authorization);
+			}
+		}
+		return Optional.empty();
+	}
+
+	private void sendTokenIntrospectionResponse(HttpServletResponse response,
+			TokenIntrospectionSuccessResponse tokenIntrospectionResponse) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		this.tokenIntrospectionHttpResponseConverter.write(tokenIntrospectionResponse, null, httpResponse);
+	}
+
+	private AbstractOAuth2Token parseJwt(AbstractOAuth2Token token) {
+		for (JwtDecoder jwtDecoder : this.jwtDecoders) {
+			try {
+				return jwtDecoder.decode(token.getTokenValue());
+			} catch (JwtException ex) {
+				// token might not be a JWT, or can't be processed with this decoder.
+			}
+		}
+		return token;
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T extends AbstractOAuth2Token> void validateToken(T token) {
+		OAuth2TokenValidator<T> tokenValidator = (token instanceof Jwt)
+				? (OAuth2TokenValidator<T>) new JwtTimestampValidator()
+				: (OAuth2TokenValidator<T>) new TokenExpirationValidator();
+		OAuth2TokenValidatorResult result = tokenValidator.validate(token);
+		if (result.hasErrors()) {
+			String errorMessages = result.getErrors().stream().map(OAuth2Error::getErrorCode)
+					.collect(Collectors.joining(", ", "Invalid Token:", "."));
+			throw new InvalidTokenException(errorMessages);
+		}
+	}
+
+	private static boolean isPrincipalAuthenticated(Authentication principal) {
```

Comment:
Note I copied this method from the _OAuth2AuthorizationEndpointFilter_ class, maybe we should move it to a utils class?